amplipi/streams.py

```diff
@@ -234,54 +255,115 @@ def connect(self, src):
 
   def disconnect(self):
+    print(f'disconnecting {self.name}!')
+    if self.mpris:
+      self.mpris.close()
+    self.mpris = None
     if self._is_running():
-      os.killpg(os.getpgid(self.proc2.pid), signal.SIGKILL)
-      self.proc.kill()
+      self.proc.stdin.close()
+      print('stopping shairport-sync')
+      self.proc.terminate()
+      if self.proc.wait(1) != 0:
+        print('killing shairport-sync')
+        self.proc.kill()
+    try:
+      subprocess.run(f'rm -r {utils.get_folder("config")}/srcs/{self.src}/*', shell=True, check=True)
+    except Exception as e:
+      print(f'Error removing airplay config files: {e}')
     self._disconnect()
-    self.proc2 = None
     self.proc = None
 
   def info(self) -> models.SourceInfo:
-    src_config_folder = f'{utils.get_folder("config")}/srcs/{self.src}'
-    loc = f'{src_config_folder}/currentSong'
-    source = models.SourceInfo(name=self.full_name(), state=self.state)
-    source.img_url = 'static/imgs/shairport.png'
+    source = models.SourceInfo(
+      name=f"Connect to {self.name} on Airplay{'2' if self.ap2 else ''}",
+      state=self.state,
+      img_url='static/imgs/shairport.png'
+    )
+
+    # if stream is airplay2 and other airplay2s exist show error message
+    if self.ap2:
+      if self.ap2_exists:
+        source.name = 'An Airplay2 stream already exists!\n Please disconnect it and try again.'
+        return source
+
+    if not self.mpris:
+      print(f'Airplay: No MPRIS object for {self.name}!')
+      return source
+
     try:
-      with open(loc, 'r') as file:
-        for line in file.readlines():
-          if line:
-            data = line.split(',,,')
-            for i in range(len(data)):
-              data[i] = data[i].strip('".')
-            source.artist = data[0]
-            source.track = data[1]
-            source.album = data[2]
-            if 'False' in data[3]:
-              source.state = 'playing'
-            else:
-              source.state = 'paused'
-            if int(data[4]):
-              source.img_url = f"/generated/shairport/srcs/{self.src}/{data[5]}"
-    except Exception as exc:
-      print(f'Failed to get currentSong - it may not exist: {exc}')
+      md = self.mpris.metadata()
+
+      if self.mpris.is_playing():
+        source.state = 'playing'
+      else:
+        # if we've been paused for a while, say we're stopped since
+        # shairport-sync doesn't really differentiate between paused and stopped
+        if source.state == 'paused' and time.time() - md.state_changed_time < self.STATE_TIMEOUT:
+          source.state = 'paused'
+        else:
+          source.state = 'stopped'
+
+      if source.state != 'stopped':
+        source.artist = md.artist
+        source.track = md.title
+        source.album = md.album
+        source.supported_cmds=list(self.supported_cmds)
+
+        if md.title != '':
+          img_name = os.listdir(f'{utils.get_folder("web")}/generated/{self.src}')[0]
```

Comment:
This throws an exception if the firectory is empty which seems like a pretty normal case